### PR TITLE
[WIP] Message contracts approvals

### PR DESCRIPTION
**Pull request by mauroservienti**
Connects to https://github.com/Particular/ServiceControl/issues/1354

### CODE REVIEW

**Comment by mauroservienti** [comment]
I'm not sure about the last commit https://github.com/Particular/ServiceControl/pull/1401/commits/4b2f980c2f4b3c40a3e54ed085295e6bb4270011

File: src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.ApproveDomainEventTypes.approved.txt [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+{
+    public class CustomCheckFailed : ServiceControl.Infrastructure.DomainEvents.IDomainEvent, ServiceControl.Infrastructure.SignalR.IUserInterfaceEvent
+    {
+        public CustomCheckFailed() { }
+        public string Category { get; set; }
+        public string CustomCheckId { get; set; }
+        public System.DateTime FailedAt { get; set; }
+        public string FailureReason { get; set; }
+        public System.Guid Id { get; set; }
+        public ServiceControl.Contracts.Operations.EndpointDetails OriginatingEndpoint { get; set; }
```
Comment (mikeminutillo) [inline src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.ApproveDomainEventTypes.approved.txt]: Hmmm. We need to find a way to find all of the child types too as they form a part of the contract. I wonder if the ApiApprovals library is the best way to do this or if some basic reflection code would be better suited to our needs here.

Comment (mikeminutillo) [reply src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.ApproveDomainEventTypes.approved.txt]: For instance, we probably don't care about interfaces or constructors for this approval.

**Comment by mikeminutillo** [comment]
> I'm not sure about the last commit 4b2f980

I wonder if we should be diffing this against the SC Contracts assembly. In fact, why aren't we just using the SC Contract assembly?

**Comment by mauroservienti** [comment]
Where is the ServiceControl.Contracts assembly?
The thing is that we want to test the public API of all the messages that we send via SignalR as well.

**Comment by mikeminutillo** [comment]
@mauroservienti SC Contracts has it's own Repo with it's own approval tests https://github.com/Particular/ServiceControl.Contracts

It's not the same set of classes though.

**Comment by mauroservienti** [comment]
@mikeminutillo I can find any approval tests for [ServiceControl.Contracts](https://github.com/Particular/ServiceControl.Contracts). Are they supposed to be in the same repo or anywhere else?

File: src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.ApproveDomainEventTypes.approved.txt (hunk 1 above)
Comment (bording) [reply src/ServiceControl.UnitTests/ApprovalFiles/APIApprovals.ApproveDomainEventTypes.approved.txt]: To clarify, the content of this file is created by PublicApiGenerator, not Particular.Approvals. If this file has unwanted stuff in it, the call to `ApiGenerator.GeneratePublicApi` is what needs to be tweaked/replaced, not `Approver.Verify`.

**Comment by mauroservienti** [comment]
@boblangley can this one be closed in favor of #1406 ?

File: src/ServiceControl.UnitTests/API/APIApprovals.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    using ServiceControl.Infrastructure.DomainEvents;
+    using ServiceControl.Infrastructure.SignalR;
+    using System.Linq;
+    using System.Runtime.CompilerServices;
+
+    [TestFixture]
+    class APIApprovals
+    {
+        [Test]
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (bording) [inline src/ServiceControl.UnitTests/API/APIApprovals.cs]: FYI These attributes aren't required when using Particular.Approvals

**Comment by boblangley** [comment]
Yes, but we can remove the attributes @bording mentioned in that PR.

**Pull request closed without merging**